packages/neo-one-editor/src/manager/PackageManager.ts

```diff
@@ -67,46 +72,103 @@ export class PackageManager {
-      depInfo.types === undefined
-        ? Promise.resolve(undefined)
-        : this.downloadDependency({ name: depInfo.name, version: depInfo.version, file: depInfo.types }),
-      this.fetchPackages(depInfo.dependencies),
+  private async fetchPackage({ name, version, types, dependencies }: DependencyInfo): Promise<Module> {
+    const [packageFilePaths, subDependencies] = await Promise.all([
+      this.getAllPackageFiles({ name, version }),
+      this.fetchPackages(dependencies),
+      this.checkTypes({ name, version, types }),
     ]);
 
+    const packageFiles = await Promise.all(
+      packageFilePaths.map(async (filePath) => ({
+        path: filePath,
+        file: await this.downloadDependency({ name, version, suffix: filePath }),
+      })),
+    );
+
     return {
-      name: depInfo.name,
-      fileName: file === undefined ? 'index.js' : file,
-      version: depInfo.version,
-      dependency: pkg,
-      packageJSON: pkgJSON,
-      types,
-      typeFile: depInfo.types === undefined ? 'index.d.ts' : depInfo.types,
-      subDependencies: subPkgs,
+      name,
+      version,
+      packageFiles,
+      subDependencies,
     };
   }
 
-  private async downloadDependency({ name, version, file }: FetchPackageInfo) {
+  private async getAllPackageFiles({ name, version, suffix }: FetchPackageInfo): Promise<ReadonlyArray<string>> {
+    const rootDirList = await this.getDirectoryList({ name, version, suffix });
+
+    const packageFiles = await Promise.all(
+      rootDirList.map(async (fileOrDir: string) => {
+        if (fileOrDir.includes('.') || fileOrDir === 'LICENSE') {
+          return fileOrDir;
+        }
+
+        return this.getAllPackageFiles({ name, version, suffix: `${fileOrDir}/` });
+      }),
+    );
+
+    return _.flatten(packageFiles);
+  }
+
+  private async getDirectoryList({ name, version, suffix }: FetchPackageInfo): Promise<ReadonlyArray<string>> {
+    const res = await fetch(this.getNPMUrl({ name, version, suffix }));
+    const htmlDir = await res.text();
+
+    return htmlDir
+      .split('\n')
+      .filter((line) => line.includes('</a></td>'))
+      .map((line) => {
+        const match = line.match(/[^>\]]+(?=<)/g);
+
+        return match === null ? '' : match[1];
+      })
+      .filter((line) => line !== '' && line !== '..')
+      .map((line) => (suffix === undefined ? line : `${suffix}${line}`));
+  }
+
+  private async checkTypes({ name, version, types }: FetchPackageInfo) {
+    if (types) {
+      return;
+    }
+    try {
+      const res = await fetch(this.getTypesUrl(name));
+      const htmlDir = await res.text();
+      if (!res.ok) {
+        return;
+      }
+
+      const typesFullName = htmlDir
+        .split('\n')
+        .filter((line) => line.includes('@types') && !line.includes('>'))[0]
+        .replace('\t', '')
+        .trim();
+
+      await this.typesCallback(name, version, typesFullName);
```

Comment:
Callback doesn't need to return anything. Also, maybe call it something like `onAddTypes`